src/CIFAR10/Reader.jl

```diff
@@ -0,0 +1,69 @@
+module Reader
+
+export readdata
+
+const NROW = 32
+const NCOL = 32
+const NCHAN = 3
+const CHUNK_SIZE = 10_000
+
+function _readdata!(io::IO,
+                    images::Array{UInt8, 4}, labels::Vector,
+                    nsrc::Integer, src_indices::AbstractVector{<:Integer},
+                    src_offset::Integer = 0, dst_indices::AbstractVector{<:Integer} = 1:size(images, 4))
+    all(ix -> 1 <= ix-src_offset <= nsrc, src_indices) ||
+        throw(ArgumentError("not all elements in parameter \"indices\" are in 1:$nsrc"))
+    img_size = map(i -> size(images, i), (1, 2, 3))
+    imagesize = prod(img_size)*sizeof(eltype(images)) + sizeof(eltype(labels))
+    if size(images, 4) == 1
+        # avoid buffer allocation for single image
+        skip(io, imagesize * (src_indices[1] - src_offset - 1))
+        labels[dst_indices[1]] = read(io, eltype(labels))
+        read!(io, images)
+    else
+        buffer = Array{UInt8}(undef, img_size)
+        pos = 0
+        @inbounds for (dst_ix, src_ix) in zip(dst_indices, src_indices)
+            nextpos = imagesize * (src_ix - src_offset - 1)
+            skip(io, nextpos - pos)
+            labels[dst_ix] = read(io, eltype(labels))
+            read!(io, buffer)
+            copyto!(view(images, :,:,:, dst_ix), buffer)
+            pos = nextpos + imagesize
+        end
+    end
+    return images, labels
+end
+
+_readdata!(file::AbstractString, images, labels, nsrc, src_indices,
+           src_offset = 0, dst_indices::AbstractVector{<:Integer} = 1:size(images, 4)) =
+    open(file, "r") do io
+        _readdata!(io, images, labels, nsrc, src_indices, src_offset, dst_indices)
+    end
```

Comment:
i think for functions like this we should use the

```
function ...
end
```

syntax

Reply:
I thought we can use a terse declaration syntax for single-instruction wrapper functions. Just tell me if you think it's critical to revert it.

Reply:
It is not critical. I can do it later